### PR TITLE
Changed all the exit and free strategy

### PR DESCRIPTION
**Pull request by dbejar-s**

### CODE REVIEW

**Comment by dbejar-s** [comment]
When type exit 42, I can't get rid of those 3 open fd at exit

image

**Comment by dbejar-s** [comment]
With redirections that throw some kind of error, I can't  get rid of the leaks (and I think I clean everything before exiting of those ones).
cat <"1""2""3""4""5"
echo <123 <456 hi | echo 42
etc...
image
image

**Comment by dbejar-s** [comment]
Había un bug y no leía la instrucción si empezaba por espacios... es decir "                                   ls"... petaba! 

Arreglado en fix_redirections con un while antes de empezar a leer
	while (ft_isspace(ins[i]))
		i++;

**Comment by pmarkaide** [comment]
The leaks can be solved by adding free_macro instead of free_ins in the validation step. I've pushed a commit to fix it

File: lib/libft/ft_strdup.c [hunk 1]
```diff
@@ -19,5 +19,6 @@ char	*ft_strdup(const char *s1)
 	copy = (char *)malloc((ft_strlen(s1) + 1) * sizeof(char));
 	if (!copy)
 		return (NULL);
-	return (ft_strcpy(copy, s1));
+	ft_memcpy(copy, s1, ft_strlen(s1) + 1);
```
Comment (pmarkaide) [inline lib/libft/ft_strdup.c:22]: Is this to fix leaks? seem a drastic solution :D

File: src/execution.c [hunk 2]
```diff
@@ -45,15 +45,16 @@ static void	execute_child_process(t_macro *macro, int index, int read_end)
 		cmd = cmd->next;
 	validation(macro, cmd);
 	if (dup_file_descriptors(macro, cmd, read_end) == -1)
-		exit(macro->exit_code);
+		exit_free(macro);
```
Comment (pmarkaide) [inline src/execution.c:48]: I think here is more correct to use exit_error to add a message

**Review by pmarkaide — COMMENTED** [review]
This looks good and is the proper solution to exit child processes cleaning all if something goes wrong before execve. Also we must add a free_macro step if execve actually fails

**Pull request merged**